client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -0,0 +1,236 @@
+        } else {
+            if (isTokenBeingRefreshed.compareAndSet(false, true)) {
+                log.debug("Resetting token");
+                resetToken(existingToken, newToken());
+            } else {
+                log.debug("Token refresh status setting to 'true' failed.");
+            }
+        }
+        String result = delegationToken.get().getValue();
+        log.trace("Exiting refreshToken()");
+        return result;
+    }
+
+    protected DelegationToken newToken() {
+        String token = Futures.getAndHandleExceptions(
+                controllerClient.getOrRefreshDelegationTokenFor(scopeName, streamName), RuntimeException::new);
+        return new DelegationToken(token, extractExpirationTime(token));
+    }
+
+    protected DelegationToken getCurrentToken() {
+        return this.delegationToken.get();
+    }
+
+    private boolean isTokenNearingExpiry() {
+        Long currentTokenExpirationTime = this.delegationToken.get().getExpiryTime();
+
+        // currentTokenExpirationTime can be null if the server returns a null delegation token
+        return currentTokenExpirationTime != null && isWithinRefreshThreshold(currentTokenExpirationTime);
+    }
+
+    private boolean isWithinRefreshThreshold(Long expirationTime) {
+        Preconditions.checkNotNull(expirationTime);
+        return isWithinRefreshThreshold(Instant.now(), Instant.ofEpochSecond(expirationTime));
+    }
+
+    @VisibleForTesting
+    boolean isWithinRefreshThreshold(Instant currentInstant, Instant expiration) {
+        return currentInstant.plusSeconds(tokenRefreshThreshold).getEpochSecond() >= expiration.getEpochSecond();
+    }
+
+    protected boolean resetToken(DelegationToken existingToken, DelegationToken newToken) {
+        boolean isTokenResetSuccessful = delegationToken.compareAndSet(existingToken, newToken);
+        if (isTokenResetSuccessful) {
+            log.debug("Successfully reset token");
+            isTokenBeingRefreshed.compareAndSet(true, false);
+        }
+        return isTokenResetSuccessful;
+    }
+
+    @VisibleForTesting
+    Long extractExpirationTime(String token) {
+        if (token == null || token.trim().equals("")) {
+            return null;
+        }
+        String[] tokenParts = token.split("\\.");
+
+        // JWT is of the format abc.def.ghe. The middle part is the body.
+        String encodedBody = tokenParts[1];
+        String decodedJsonBody = new String(Base64.getDecoder().decode(encodedBody));
+
+        return parseExpirationTime(decodedJsonBody);
+    }
+
+    @VisibleForTesting
+    Long parseExpirationTime(String json) {
+        Long result = null;
+        if (json != null && !json.trim().equals("")) {
+            // Sample inputs:
+            //    {"sub":"subject","aud":"segmentstore","iat":1569837384,"exp":1569837434}
+            //    {"sub": "subject","aud": "segmentstore","iat": 1569837384,"exp": 1569837434}
+            Pattern pattern = Pattern.compile("\"exp\":\\s?(\\d+)");
```

Comment:
Do you need to compile the pattern every time you invoke this method? You can stash this in a private static final field.

Reply:
Yes, it makes sense to cache the pattern in a private static final field. I've now modified it.